src/Servers.jl

```diff
@@ -410,108 +118,110 @@ function handle_connection(f::Function, c::Connection;
+Base.close(s::Server2) = close(s.server)
+
+function listenloop(h::Handler, server,
+    tcpisvalid=x->true, connectioncounter=Ref(0),
+    reuse_limit::Int=1, readtimeout::Int=0, verbose::Bool=false)
+    count = 1
+    while isopen(server)
+        try
+            io = accept(server)
+            io === nothing && continue
+            if !tcpisvalid(io)
+                verbose && @info "Accept-Reject:  $io"
+                close(io)
+                continue
+            end
+            connectioncounter[] += 1
+            conn = Connection(server.hostname, server.hostport, 0, 0, true, io)
+            @async begin
+                try
+                    verbose && @info "Accept ($count):  $conn"
+                    handle(h, conn)
+                    verbose && @info "Closed ($count):  $conn"
+                catch e
+                    @error exception=(e, stacktrace(catch_backtrace()))
+                finally
+                    connectioncounter[] -= 1
+                    close(io)
+                    verbose && @info "Closed ($count):  $conn"
+                end
+            end
+        catch e
+            if e isa InterruptException
+                @warn "Interrupted: listen($server)"
+                close(server)
+                break
+            end
             @error exception=(e, stacktrace(catch_backtrace()))
         end
-        close(t)
-    finally
-        final_transaction && close(t.c)
+        count += 1
     end
     return
 end
 
-"""
-Execute stream processing function `f`.
-If there is an error and the stream is still open,
-send a 500 response with the error message.
+getinet(host::String, port::Integer) = Sockets.InetAddr(parse(IPAddr, host), port)
+getinet(host::IPAddr, port::Integer) = Sockets.InetAddr(host, port)
 
-Close the `Stream` for read and write (in case `f` has not already done so).
-"""
-function handle_stream(f::Function, http::Stream)
+function listen(h::Handler, host::Union{IPAddr, String}, port::Integer;
+    tcpref::Union{Ref, Nothing}=nothing,
+    reuseaddr::Bool=false,
+    sslconfig::Union{MbedTLS.SSLConfig, Nothing}=nothing,
+    tcpisvalid::Union{Function, Nothing}=nothing,
+    ratelimit::Union{Rational{Int}, Nothing}=nothing,
+    connectioncounter::Ref{Int}=Ref(0),
+    reuse_limit::Int=1, readtimeout::Int=0, verbose::Bool=false)
 
-    try
-        if applicable(f, http)
-            f(http)
-        else
-            handle_request(f, http)
-        end
-    catch e
-        if isopen(http) && !iswritable(http)
-            @error exception=(e, stacktrace(catch_backtrace()))
-            http.message.response.status = 500
-            startwrite(http)
-            write(http, sprint(showerror, e))
+    inet = getinet(host, port)
+    if tcpref !== nothing
+        tcpserver = tcpref[]
+    elseif reuseaddr
+        tcpserver = Sockets.TCPServer(; delay=false)
+        if Sys.islinux() || Sys.isapple()
```

Comment:
Do you want  to be using `Sys.isunix` ? Or is there a reason to exclude BSD?

Reply:
Ah, good call.